external_snaps_tests/__main__.py

```diff
@@ -97,9 +97,13 @@ def _bzr_branch(url):
 
 def _build_snaps(path, cleanbuild=False, keep_dir=False):
     try:
-        for dirpath, _, filenames in os.walk(path):
-            if 'snapcraft.yaml' in filenames or '.snapcraft.yaml' in filenames:
+        for dirpath, dirnames, filenames in os.walk(path, topdown=True):
+            if ('snap' in dirnames
+                    or 'snapcraft.yaml' in filenames
+                    or '.snapcraft.yaml' in filenames):
                 _build_snap(dirpath, cleanbuild, keep_dir)
+                # Do not recurse in any directory.
+                dirnames = []
```

Comment:
I am. I think. I don't really understand your question.
The docs say that recursion will happen only on the directories that remain in dirnames. If dirnames is empty, no recursion will happen.